Userland/Games/ColorLines/MarbleBoard.h

```diff
@@ -0,0 +1,356 @@
+
+        constexpr Point connected_four_ways[4] = {
+            { 0, -1 }, // to the top
+            { 0, 1 },  // to the bottom
+            { -1, 0 }, // to the left
+            { 1, 0 }   // to the right
+        };
+
+        while (!queue.is_empty()) {
+            auto current = queue.dequeue();
+            if (current == to) {
+                while (current != from) {
+                    path.add_point(current);
+                    for (auto delta : connected_four_ways)
+                        if (auto next = current.translated(delta); MarbleBoard::in_bounds(next) && trace[next] < trace[current]) {
+                            current = next;
+                            break;
+                        }
+                }
+                path.add_point(current);
+                return true;
+            }
+            for (auto delta : connected_four_ways)
+                add_path_point(current.translated(delta), trace[current] + 1);
+        }
+        return false;
+    }
+
+    [[nodiscard]] bool check_and_remove_marbles()
+    {
+        m_removed_marbles.clear();
+        constexpr Point connected_four_ways[] = {
+            { -1, 0 },  // to the left
+            { 0, -1 },  // to the top
+            { -1, -1 }, // to the top-left
+            { 1, -1 }   // to the top-right
+        };
+        HashTable<Marble, Traits<Marble>> marbles;
+        for_each_cell([&](Point current_point) {
+            if (is_empty_cell_at(current_point))
+                return IterationDecision::Continue;
+            auto const color { color_at(current_point) };
+            for (auto direction : connected_four_ways) {
+                size_t marble_count = 0;
+                for (auto p = current_point; in_bounds(p) && color_at(p) == color; p.translate_by(direction))
+                    ++marble_count;
+                if (marble_count >= number_of_marbles_to_remove)
+                    for (auto p = current_point; in_bounds(p) && color_at(p) == color; p.translate_by(direction))
+                        marbles.set({ p, color });
+            }
+            return IterationDecision::Continue;
+        });
+        m_removed_marbles.ensure_capacity(marbles.size());
+        for (auto const& marble : marbles) {
+            m_removed_marbles.append(marble);
+            clear_color_at(marble.position());
+        }
+        return !m_removed_marbles.is_empty();
+    }
+
+    [[nodiscard]] PreviewMarbles const& preview_marbles() const
+    {
+        return m_preview_marbles;
+    }
+
+    [[nodiscard]] bool has_selected_marble() const
+    {
+        return m_selected_marble != nullptr;
+    }
+
+    [[nodiscard]] SelectedMarble const& selected_marble() const
+    {
+        VERIFY(has_selected_marble());
+        return *m_selected_marble;
+    }
+
+    [[nodiscard]] bool select_marble(Point point)
+    {
+        if (!is_empty_cell_at(point)) {
+            m_selected_marble = make<SelectedMarble>(point, color_at(point));
```

Comment:
Keeping a single SelectedMarble around would avoid repeatedly allocating/discarding one.

Reply:
Here, the `OwnPtr<SelectedMarble>` is implementing the `Maybe` or `optional` pattern. I.e. the marble may be selected, and if it is selected it stores the location/color. It is at most one selected marble during the game, and I think it is ok as it is.